src/server/dfly_main.cc

```diff
@@ -96,6 +99,114 @@ namespace dfly {
 
 namespace {
 
+using util::http::TlsClient;
+
+const std::string_view kInvalidVersion{"unknown version"};
+
+std::string GetVersionString(const std::string& from) {
+  // The server sends a message such as {"latest": "0.12.0"}
+  const auto reg_match_expr = R"(\{\"latest"\:[ \t]*\"([0-9]+\.[0-9]+\.[0-9]+)\"\})";
+  VLOG(1) << "checking version '" << from << "'";
+  auto const regex = std::regex(reg_match_expr);
+  std::smatch match;
+  if (std::regex_match(from, match, regex) && match.size() > 1) {
+    // the second entry is the match to the group that holds the version string
+    return match[1].str();
+  } else {
+    return std::string{kInvalidVersion};
+  }
+}
+
+std::string GetRemoteVersion(ProactorBase* proactor, SSL_CTX* ssl_context, const std::string host,
+                             std::string_view service, const std::string& resource,
+                             const std::string& ver_header) {
+  namespace bh = boost::beast::http;
+  using ResponseType = bh::response<bh::string_body>;
+
+  bh::request<bh::string_body> req{bh::verb::get, resource, 11 /*http 1.1*/};
+  req.set(bh::field::host, host);
+  req.set(bh::field::user_agent, ver_header);
+  ResponseType res;
+  TlsClient http_client{proactor};
+  http_client.set_connect_timeout_ms(2000);
+
+  auto ec = http_client.Connect(host, service, ssl_context);
+  if (!ec) {
+    ec = http_client.Send(req, &res);
+    if (!ec) {
+      VLOG(1) << "successfully got response from HTTP GET for host " << host << ":" << service
+              << "/" << resource << " response code is " << res.result();
+
+      if (res.result() == bh::status::ok) {
+        return GetVersionString(res.body());
+      }
+    } else {
+      VLOG(1) << "failed to process send HTTP GET to " << host << "@" << service << " at resource '"
+              << resource << "'";
+    }
+  } else {
+    VLOG(1) << "failed to connect: " << ec.message();
+  }
+  return std::string{kInvalidVersion};
+}
+
+struct VersionMonitor {
+  fibers_ext::Fiber version_fiber_;
+  fibers_ext::Done monitor_ver_done_;
+
+  void Run(ProactorPool* proactor_pool) {
+    version_fiber_ = proactor_pool->GetNextProactor()->LaunchFiber(
+        [this, pb = proactor_pool->GetNextProactor()] { RunTask(pb); });
```

Comment:
you do not need `pb  = proactor_pool->GetNextProactor()`. You can just use `ProactorBase::me()` to access your thread-local proactor within a fiber inside `RunTask`

Reply:
I was wondering how can I get the local proactor - thanks